### PR TITLE
fix bugs of parsing network topology in text generation task

### LINKED ISSUE #2349: StaticInputV2 and GeneratedInputV2 missing in paddle.v2.layer
From the [MT chapter in the DL101 book](https://github.com/PaddlePaddle/book/blob/develop/08.machine_translation/train.py) and [the machine translation demo](https://github.com/PaddlePaddle/Paddle/blob/develop/demo/seqToseq/api_train_v2.py) with the new python API, `paddle.layer.StaticInputV2` and `paddle.layer.GeneratedInputV2` throws an AttributeError, e.g.:

```python
AttributeError                            Traceback (most recent call last)
<ipython-input-5-5ce86945bbbe> in <module>()
----> 1 cost = seqToseq_net(source_dict_dim, target_dict_dim)

/Users/liling.tan/seqtoseq.py in seqToseq_net(source_dict_dim, target_dict_dim, is_generating)
    160 
    161     decoder_group_name = "decoder_group"
--> 162     group_input1 = paddle.layer.StaticInputV2(input=encoded_vector, is_seq=True)
    163     group_input2 = StaticInputV2(input=encoded_proj, is_seq=True)
    164     group_inputs = [group_input1, group_input2]

AttributeError: 'module' object has no attribute 'StaticInputV2'
```

A closer look, they're missing from the top level `paddle.layer.__init__.py`:

```python
>>> import paddle.v2 as paddle
>>> dir(paddle.layer)
['AggregateLevel', 'BaseGeneratedInput', 'ExpandLevel', 'GeneratedInput', 'LayerOutput', 'ModelConfig', 'SubModelConfig', 'SubsequenceInput', '__all__', '__builtins__', '__convert_name__', '__convert_to_v2__', '__data_layer__', '__doc__', '__file__', '__get_used_evaluators__', '__get_used_layers__', '__get_used_parameters__', '__get_used_submodels__', '__map_data_docstr__', '__name__', '__need_to_keep__', '__need_to_wrap__', '__package__', '__trim_submodel__', 'addto', 'batch_norm', 'beam_search', 'bilinear_interp', 'block_expand', 'classification_cost', 'collections', 'concat', 'config_base', 'context_projection', 'conv_operator', 'conv_projection', 'conv_shift', 'convex_comb', 'copy', 'cos_sim', 'cp', 'crf', 'crf_decoding', 'cross_channel_norm', 'cross_entropy_cost', 'cross_entropy_with_selfnorm_cost', 'ctc', 'data', 'dotmul_operator', 'dotmul_projection', 'embedding', 'eos', 'expand', 'fc', 'first_seq', 'full_matrix_projection', 'get_layer', 'get_output', 'gru_step', 'gru_step_naive', 'grumemory', 'hsigmoid', 'huber_cost', 'identity_projection', 'img_cmrnorm', 'img_conv', 'img_pool', 'interpolation', 'lambda_cost', 'last_seq', 'linear_comb', 'lstm_step', 'lstmemory', 'max_id', 'maxout', 'memory', 'mixed', 'mse_cost', 'multi_binary_label_cross_entropy_cost', 'multiplex', 'name', 'nce', 'new_name', 'obj', 'out_prod', 'pad', 'parse_network', 'pooling', 'power', 'print', 'priorbox', 'rank_cost', 're', 'recurrent', 'recurrent_group', 'regression_cost', 'repeat', 'rotate', 'sampling_id', 'scaling', 'scaling_projection', 'selective_fc', 'seq_concat', 'seq_reshape', 'slope_intercept', 'smooth_l1_cost', 'spp', 'sum_cost', 'sum_to_one_norm', 'table_projection', 'tensor', 'trans', 'trans_full_matrix_projection', 'v1_layers', 'warp_ctc']

>>> 'StaticInputV2' in dir(paddle.layer)
False
>>> 'GeneratedInputV2' in dir(paddle.layer)
False
```

Currently the ad-hoc solution is to use the old `paddle.trainer_config_helpers`, e.g.

```
>>> from paddle.trainer_config_helpers import GeneratedInput 
>>> from paddle.trainer_config_helpers import StaticInput 
```

----

This is from version 0.10.0 of the Python API.

```
$ pip show paddle
Name: paddle
Version: 0.10.0
Summary: Parallel Distributed Deep Learning
Home-page: UNKNOWN
Author: UNKNOWN
Author-email: UNKNOWN
License: UNKNOWN
Location: /usr/local/lib/python2.7/site-packages
Requires: rarfile, matplotlib, protobuf, opencv-python, numpy, requests


$ paddle version
PaddlePaddle 0.10.0, compiled with
    with_avx: ON
    with_gpu: OFF
    with_double: OFF
    with_python: ON
    with_rdma: OFF
    with_timer: OFF
```

### LINKED ISSUE #2385: beam search cannot be parsed correctly.
- This PR https://github.com/PaddlePaddle/Paddle/pull/2288 changes the way v2 APIs parse the network configuration. But currently, `beam search` cannot be parsed correctly. It causes bugs to machine translation in PaddleBook.

- This PR https://github.com/PaddlePaddle/Paddle/pull/2384 fix the training process but hasn't fixed bugs in the generation.

- I change `is_generating=True` in this line https://github.com/PaddlePaddle/book/blob/develop/08.machine_translation/train.py#L129, the error information go as follows
```text
I0605 18:50:24.263185 18007 Util.cpp:166] commandline:  --use_gpu=False --trainer_count=1
F0605 18:50:46.687893 18007 NeuralNetwork.h:94] Check failed: it != layerMap_.end() Unknown layer __decoder_group_eos_layer__@decoder_group
*** Check failure stack trace: ***
    @     0x7f7019fed7bd  google::LogMessage::Fail()
    @     0x7f7019ff126c  google::LogMessage::SendToLog()
    @     0x7f7019fed2e3  google::LogMessage::Flush()
    @     0x7f7019ff277e  google::LogMessageFatal::~LogMessageFatal()
    @     0x7f7019c2afef  paddle::NeuralNetwork::getLayer()
    @     0x7f7019c3932c  paddle::RecurrentGradientMachine::resizeOrCreateFrames()
    @     0x7f7019c3873b  paddle::RecurrentGradientMachine::init()
    @     0x7f7019bd3be9  paddle::RecurrentLayerGroup::initSubNetwork()
    @     0x7f7019c285c5  paddle::NeuralNetwork::init()
    @     0x7f7019c52bde  paddle::GradientMachine::create()
    @     0x7f7019fc10cc  GradientMachine::createFromPaddleModelPtr()
    @     0x7f7019fc120d  GradientMachine::createByConfigProtoStr()
    @     0x7f7019ab07d5  _wrap_GradientMachine_createByConfigProtoStr__SWIG_0
    @     0x7f7019ab0e10  _wrap_GradientMachine_createByConfigProtoStr
    @     0x7f7021f513a3  PyEval_EvalFrameEx
    @     0x7f7021f53130  PyEval_EvalCodeEx
    @     0x7f7021f514a1  PyEval_EvalFrameEx
    @     0x7f7021f53130  PyEval_EvalCodeEx
    @     0x7f7021f514a1  PyEval_EvalFrameEx
    @     0x7f7021f53130  PyEval_EvalCodeEx
    @     0x7f7021edf27d  function_call
    @     0x7f7021eb70f3  PyObject_Call
    @     0x7f7021ec9f7f  instancemethod_call
    @     0x7f7021eb70f3  PyObject_Call
    @     0x7f7021f0d80e  slot_tp_init
    @     0x7f7021f0c478  type_call
    @     0x7f7021eb70f3  PyObject_Call
    @     0x7f7021f50887  PyEval_EvalFrameEx
    @     0x7f7021f53130  PyEval_EvalCodeEx
    @     0x7f7021f514a1  PyEval_EvalFrameEx
    @     0x7f7021f51c56  PyEval_EvalFrameEx
    @     0x7f7021f53130  PyEval_EvalCodeEx
```
~~Besides `eos_layer`, several other layers are not created in generation.~~

### PR DESCRIPTION
**Pull request by lcy-seso**
fix #2349. fix #2385

1. fix the bug that `eos_layer` is not trimmed as a used layer in text generation.
2. fix a bug in parsing the extra evaluator in SRL demo.
3. move `dropout_layer` from `paddle.network` to `paddle.layer`
4. export `StaticInput` and `GeneratedInput`. Currently, `StaticInputV2` and `GeneratedInputV2` are not available anymore, use `StaticInput` and `GeneratedInput` instead.

### CODE REVIEW

**Comment by lcy-seso** [comment]
Do not approve. `beam search` still cannot be parsed correctly in v2 API.

File: doc/api/v2/config/layer.rst [hunk 1]
```diff
@@ -434,10 +434,19 @@ smooth_l1_cost
...
 
 eos
 ---
 ..  autoclass:: paddle.v2.layer.eos
     :noindex:
+
+Miscs
+=====
+
+dropout_layer
```
Comment (luotao1) [inline doc/api/v2/config/layer.rst]: v2 api中默认后面不加layer，dropout_layer后面为什么要加呢

Comment (lcy-seso) [inline doc/api/v2/config/layer.rst]: 已经修改。将 drop 从 network 下移动到 layer 下，潜在会影响一些旧的配置。

**Comment by helinwang** [comment]
Please ignore "TeamCity build failed", see: https://github.com/PaddlePaddle/Paddle/issues/2401

**Comment by lcy-seso** [comment]
@helinwang I get it. Thank you.

File: python/paddle/v2/layer.py [hunk 2]
```diff
@@ -235,11 +235,10 @@ def __trim_submodel__(old_submodel, layer_names, input_layer_names,
 def parse_network(output_layers, extra_layers=None):
     if not isinstance(output_layers, collections.Sequence):
         output_layers = [output_layers]
-    if extra_layers is not None and not isinstance(extra_layers,
-                                                   collections.Sequence):
```
Comment (lcy-seso) [inline python/paddle/v2/layer.py]: - Topology 的构造函数中会对 extra_layers 的类型进行检查
- 检查的过程中，如果传入的extra_layer 不是 python list，会转化为 list
https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/v2/topology.py#L43
- 代码走到这里时 extra_layer 一定是一个 python list，这一行条件判断的后一个条件会将 extra_layer 清空

File: python/paddle/v2/layer.py [hunk 3]
```diff
@@ -248,18 +255,21 @@ def parse_network(output_layers, extra_layers=None):
...
+        model_config.output_layer_names.append(layer.full_name)
+        output_layer_names.add(layer.full_name)
+
     for l in cp.g_config.model_config.layers:
         if l.name not in layer_names:
             continue
         model_config.layers.extend([l])
         if l.type == 'data':
+            if l.name in model_config.output_layer_names:
+                continue
```
Comment (emailweixu) [inline python/paddle/v2/layer.py]: Why is this needed?

File: python/paddle/v2/topology.py [hunk 4]
```diff
@@ -91,8 +90,9 @@ def data_type(self):
         [('image', dense_vector(768)), ('label', integer_value(10))]
         """
         data_layers = self.data_layers()
+
         return [(nm, data_layers[nm].data_type)
-                for nm in self.proto().input_layer_names]
+                for nm in self.proto().input_layer_names if nm in data_layers]
```
Comment (emailweixu) [inline python/paddle/v2/topology.py]: when is input_layer not a data_layer?

Comment (lcy-seso) [inline python/paddle/v2/topology.py]: This is a mistake. I have fixed this.

File: python/paddle/v2/layer.py (hunk 3 above)
Comment (emailweixu) [inline python/paddle/v2/layer.py]: Need to add a comment. Otherwise it's very hard to understand.

Comment (lcy-seso) [inline python/paddle/v2/layer.py]: done.
